### PR TITLE
NEW Give feedback of password strength

### PR DESCRIPTION
**Pull request by emteknetnz**
~~Issue https://github.com/silverstripe/silverstripe-framework/issues/11413~~

~~This PR is part of a POC to give UX feedback of the new password~~

Adds functionality to provide feedback to the user about the strength of their password.

See other PRs in the linked issue (e.g. admin PR includes relevant JS and CSS)

## Issue

- https://github.com/silverstripe/silverstripe-framework/issues/11475

### CODE REVIEW

**Comment by GuySartorelli** [comment]
You might want to add a bit of a description about what you're actually doing here, because it's clearly more than just changing the default from "strong" to "medium"

**Comment by emteknetnz** [comment]
Updated description, this PR is just a POC

**Comment by christopherdarling** [comment]
Many users won't really understand what medium / strong passwords are and maybe what's missing to meet the medium strength. Is it possible to include in the notice some feedback of what is missing? or maybe if easier, what the medium strength criteria is?

**Comment by GuySartorelli** [comment]
The criteria is entropy based for the default password validator in CMS 6, so it's not a matter of meeting specific rules like "must have at least 15 characters total" or "must have at least one number" - and since we're using symfony's validation component for it they could change the specific implementation details at any time. Any specific details we give would be verbose and prone to becoming outdated.

File: src/Dev/TestSession.php [hunk 1]
```diff
@@ -15,6 +15,7 @@
 use SilverStripe\Core\Extensible;
 use SilverStripe\Core\Injector\Injector;
 use Symfony\Component\DomCrawler\Crawler;
+use Symfony\Component\DomCrawler\Field\FormField;
```
Comment (GuySartorelli) [inline src/Dev/TestSession.php]: Why is this being added?

File: src/Forms/ConfirmedPasswordField.php [hunk 2]
```diff
@@ -708,12 +781,33 @@ public function getRequireStrongPassword(): bool
...
+        return $this;
+    }
+
+    /**
+     * Get whether the field is on a member form
+     */
+    public function getIsOnMemberForm(): bool
+    {
+        return $this->isOnMemberForm;
+    }
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]: Why does this need to be a public getter?

File: tests/php/Forms/ConfirmedPasswordFieldTest.php [hunk 3]
```diff
@@ -482,4 +500,227 @@ public function testChildFieldsAreRequired(bool $canBeEmpty, bool $required, boo
...
+                'requireStrongPassowrd' => false,
+                'validatorClass' => EntropyPasswordValidator::class,
+                'passwordStrengthCall' => null,
+                'passwordStrengthConfig' => PasswordStrength::STRENGTH_WEAK,
+                'expectedDataStrengthUrl' => 'field/Test/strength',
+                'expectedMinStrength' => '1',
+            ],
+            'min-strength-weak-explicit-very-strong' => [
+                'isOnMemberForm' => true,
+                'requireStrongPassowrd' => false,
```
Comment (GuySartorelli) [inline tests/php/Forms/ConfirmedPasswordFieldTest.php:681]: There should be a test for both of these being set to `true` and having a different min strength on the field than in the validator, so we can be sure that combination works. It's an edge case but it's valid.

File: src/Forms/ConfirmedPasswordField.php [hunk 4]
```diff
@@ -753,4 +847,36 @@ private function getAttributesHTMLForChild(FormField $child): DBField
...
+    private function canEvaluateStrength(): bool
+    {
+        if (!$this->getIsOnMemberForm() && !$this->getRequireStrongPassword()) {
+            return false;
+        }
+        $validator = $this->getApplicableValidator();
+        if ($validator) {
+            return $validator->canEvaluateStrength();
+        }
+        return false;
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]:
```suggestion
    if ($this->getRequireStrongPassword()) {
        return true;
    }
    return $this->getIsOnMemberForm() && is_a(Member::password_validator(), EntropyPasswordValidator::class);
```
We've agreed that only `EntropyPasswordValidator` is relevant for member forms, so we can evaluate strength if:
1. We require a strong password on this field explicitly (regardless of whether this is on a member form or not), or
2. we're on a member form and the member is using an `EntropyPasswordValidator`.

See also https://github.com/silverstripe/silverstripe-framework/pull/11417#discussion_r2160516481 which suggests an alternative approach to this method

File: src/Forms/ConfirmedPasswordField.php [hunk 5]
```diff
@@ -753,4 +847,36 @@ private function getAttributesHTMLForChild(FormField $child): DBField
...
+            return Member::password_validator();
+        }
+        $spec = Injector::inst()->getServiceSpec(PasswordValidator::class);
+        // Do not attempt to instantiate abstract class PasswordValidator
+        // This can happen during unit tests when the member password validator is set to null
+        if ($spec && $spec['class'] !== PasswordValidator::class) {
+            return PasswordValidator::singleton();
+        }
+        return null;
+    }
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]: I don't think this method is necessary. We should be getting Symfony to tell us the password strength, rather than asking one of our own password validators.
See https://github.com/silverstripe/silverstripe-framework/pull/11417#discussion_r2160514601

File: src/Forms/ConfirmedPasswordField.php [hunk 6]
```diff
@@ -708,12 +781,33 @@ public function getRequireStrongPassword(): bool
...
+        return $this->isOnMemberForm;
+    }
+
     /**
      * Set minimum password strength. Only applies if requireStrongPassword is true
      * See https://symfony.com/doc/current/reference/constraints/PasswordStrength.html#minscore
      */
     public function setMinPasswordStrength(int $strength): static
     {
+        $this->minPasswordStrengthExplicitlySet = true;
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]: What is this new boolean for? Probably not needed, see https://github.com/silverstripe/silverstripe-framework/pull/11417#discussion_r2160514601

File: src/Forms/ConfirmedPasswordField.php [hunk 7]
```diff
@@ -185,6 +248,16 @@ public function Field($properties = [])
...
             }
+            if (str_ends_with($field->getName(), '[_Password]') && $this->canEvaluateStrength()) {
+                if (!$this->minPasswordStrengthExplicitlySet) {
+                    $validator = $this->getApplicableValidator();
+                    if (is_a($validator, EntropyPasswordValidator::class)) {
+                        $this->setMinPasswordStrength($validator->getRequiredStrength());
+                    }
+                }
+                $field->setAttribute('data-strength-url', $this->Link('strength'));
+                $field->setAttribute('data-min-strength', $this->getMinPasswordStrength());
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]: Why are we calling `$this->setMinPasswordStrength()`? Isn't this value only needed for the `data-min-strength` attribute?

```suggestion
                $minStrength = $this->getRequireStrongPassword() ? $this->getMinPasswordStrength() : 0;
                if ($this->getIsOnMemberForm()) {
                    $validator = Member::password_validator();
                    if (is_a($validator, EntropyPasswordValidator::class)) {
                        $minStrength = max($minStrength, $validator->getRequiredStrength());
                    }
                }
                $field->setAttribute('data-strength-url', $this->Link('strength'));
                $field->setAttribute('data-min-strength', $minStrength);
```

Set the attribute to the highest of:
1. `$this->getMinPasswordStrength()` (if `requireStrongPassword` is true)
2. The required strength of `EntropyPasswordValidator` (if we're on a member form and that's the current validator)
3. 0 (which shouldn't be used anyway because `canEvaluateStrength()` already returned true)

Comment (GuySartorelli) [reply src/Forms/ConfirmedPasswordField.php]: A shortcut to all this would be to replace `canEvaluateStrength()` with `getMinPasswordStrengthForEvaluation()` that returns the `$minStrength` above, and returns `-1` if we shouldn't be evaluating strength in the first place. We then get both the boolean and the strength value in a single return value from a single method (which we're only calling here anyway)

So something like

```php
public function getMinPasswordStrengthForEvaluation(): int
{
    $minStrength = $this->getRequireStrongPassword() ? $this->getMinPasswordStrength() : -1;
    if ($this->getIsOnMemberForm()) {
        $validator = Member::password_validator();
        if (is_a($validator, EntropyPasswordValidator::class)) {
            $minStrength = max($minStrength, $validator->getRequiredStrength());
        }
    }
    return $minStrength;
}
```

And then this method can be updated to

```php
$minStrength = $this->getMinPasswordStrengthForEvaluation();
if ($minStrength >= 0 && str_ends_with($field->getName(), '[_Password]')) {
      $field->setAttribute('data-strength-url', $this->Link('strength'));
      $field->setAttribute('data-min-strength', $minStrength);
}
```

File: src/Forms/ConfirmedPasswordField.php [hunk 8]
```diff
@@ -155,6 +181,43 @@ public function __construct(
...
+        $response = HTTPResponse::create();
+        $json = json_decode($request->getBody() ?? '', true);
+        if (!$json || !array_key_exists('password', $json) || !$request->isPOST()) {
+            $response->setStatusCode(400);
+            return $response;
+        }
+        $password = $json['password'];
+        $validator = PasswordValidator::create();
+        $strength = $validator->evaluateStrength($password);
+        $strengthLabel = $validator->getStrengthLabel($strength);
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]:
```suggestion
        $strength = PasswordStrengthValidator::estimateStrength($password);
        $strengthLabel = $validator->getStrengthLabel($strength);
```

[Symfony gives us this funcitonality](https://symfony.com/doc/current/reference/constraints/PasswordStrength.html#customizing-the-password-strength-estimation).

We should also move the `getStrengthLabel()` either into this class or specifically only in `EntropyPasswordValidator` (but probably here since this is where we're using it). We don't have to implement any of this in our `PasswordValidator` or `EntropyPasswordValidator` classes.

File: src/Forms/ConfirmedPasswordField.php [hunk 9]
```diff
@@ -37,6 +47,11 @@ class ConfirmedPasswordField extends FormField
      */
     public int $maxLength = 0;
 
+    /**
+     * Whether the field is on a member form
+     */
+    private bool $isOnMemberForm = false;
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php:53]: The edit form in the CMS (both in `admin/security` and in `admin/myprofile`) already has a reference to the member record which we can fetch from the field with `$this->getForm()->getRecord()`.

If we set the member record into the change password form we could remove this `isOnMemberForm` property and its setter method, and instead check `is_a($this->getForm()?->getRecord(), Member::class)`.

I'm not terribly fussed about this so feel free to keep as is if you prefer this setter pattern.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
This review was focused on the implementation in `ConfirmedPasswordField` specifically, I haven't looked closely at any of the rest of it at this stage.

File: src/Forms/ConfirmedPasswordField.php (hunk 4 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php:880]: > We've agreed that only EntropyPasswordValidator is relevant for member forms,

I'm not sure what you're referring to here

`Member::set_password_validator(?PasswordValidator $validator = null)` allows setting a non-EntropyPasswordValidator so we need too account for that?

File: src/Forms/ConfirmedPasswordField.php (hunk 9 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php:53]: Will keep as is as recommendation doesn't appear to work, debugging at ConfirmedPasswordField::canEvaluateStrength() (called by ConfirmedPasswordField::Field() so at a late point where rendering html)

$this->getForm()
// SilverStripe\Security\MemberAuthenticator\ChangePasswordForm

$this->getForm()->getRecord()
// null

Comment (GuySartorelli) [reply src/Forms/ConfirmedPasswordField.php:53]: It will work if you do what I suggested
> If we set the member record into the change password form

As I said, the form _in the CMS_ has the record set. The change password form _does not_ so to use this implementation you'd have to set the record in the form.

File: src/Forms/ConfirmedPasswordField.php (hunk 2 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: Removed

File: src/Dev/TestSession.php (hunk 1 above)
Comment (emteknetnz) [reply src/Dev/TestSession.php]: Removed

File: src/Forms/ConfirmedPasswordField.php (hunk 9 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php:53]: Opps sorry missed that

Still won't work, in the "lost password flow" when you click an email link with the security hash `Security::getCurrentUser()` will return null i.e. we could never call `$form->loadDataFrom($member)`

We could probably add some more API somewhere else to get the current member from the has (it's retrieved in ChangePasswordHandler after all) though that wouldn't account for the MFA reset account flow

Better to just keep the new API where it is IMO

Comment (GuySartorelli) [reply src/Forms/ConfirmedPasswordField.php:53]: We know who the hash belongs to though so we can get the member record and attach it to the form. Probably need a new public `setRecord()` method on form.
It will work if you want to do it, and I can help if you want to explore it. If you'd rather not, that's fine too.

File: src/Forms/ConfirmedPasswordField.php (hunk 6 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: Removed

File: src/Forms/ConfirmedPasswordField.php (hunk 5 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: I've kept it, it's still useful as the relevant validator changes depending if we're on a member for or not. Also useful to prevent attempting to instantiate an abstract class during unit tests

File: src/Forms/ConfirmedPasswordField.php (hunk 4 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: I've removed the whole method

File: tests/php/Forms/ConfirmedPasswordFieldTest.php (hunk 3 above)
Comment (emteknetnz) [reply tests/php/Forms/ConfirmedPasswordFieldTest.php:681]: Added

Note I've added another test to explicitly check that calls to setMinPasswordStrength() are ignored when requireStrongPassword is false and isOnMemberForm is true

I think this is correct because the docblock for setMinPasswordStrength() says `Only applies if requireStrongPassword is true`

File: src/Forms/ConfirmedPasswordField.php (hunk 8 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: Updated

File: src/Forms/ConfirmedPasswordField.php (hunk 7 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: Updated

**Comment by GuySartorelli** [comment]
CI is failing

File: src/Forms/ConfirmedPasswordField.php [hunk 10]
```diff
@@ -155,6 +177,47 @@ public function __construct(
...
+        if (!$json || !array_key_exists('password', $json) || !$request->isPOST()) {
+            $response->setStatusCode(400);
+            return $response;
+        }
+        $password = $json['password'];
+        $strength = -1;
+        $strengthLabel = '';
+        $validator = $this->getApplicableValidator();
+        if (is_a($validator, EntropyPasswordValidator::class)) {
+            $strength = PasswordStrengthValidator::estimateStrength($password);
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]:
```suggestion
        if ($this->getMinPasswordStrengthForEvaluation() >= 0) {
            $strength = PasswordStrengthValidator::estimateStrength($password);
```
We don't actually need a member `PasswordValidator` at all. All of the logic for this condition is already in `getMinPasswordStrengthForEvaluation()` - i.e. one of:

1. `requireStrongPassword()` is true
1. we're on a member form and the current member validator is a `EntropyPasswordValidator`

File: src/Forms/ConfirmedPasswordField.php [hunk 11]
```diff
@@ -185,7 +248,16 @@ public function Field($properties = [])
                     $field->setAttribute($name, $value);
                 }
             }
-
+            if (str_ends_with($field->getName(), '[_Password]')) {
+                $validator = $this->getApplicableValidator();
+                if (is_a($validator, EntropyPasswordValidator::class)) {
+                    $minStrength = $this->getMinPasswordStrengthForEvaluation();
+                    if ($minStrength !== -1) {
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]:
```suggestion
                $minStrength = $this->getMinPasswordStrengthForEvaluation();
                if ($minStrength >= 0) {
```
The logic is all in `getMinPasswordStrengthForEvaluation()` - we don't need an "applicable validator".

File: src/Forms/ConfirmedPasswordField.php [hunk 12]
```diff
@@ -708,6 +780,18 @@ public function getRequireStrongPassword(): bool
         return $this->requireStrongPassword;
     }
 
+    /**
+     * Set whether the field is on a Member form
+     * If this is set to true then setRequireStrongPassword() should be set to false
+     * Thisi is because we do not want to use this fields validate() method to validate
+     * the password, instead we want to validate using MemberAuthenticator
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]:
```suggestion
    /**
     * Set whether the field is on a Member form. This is used to determine whether and how to evaluate password strength for this form field.
```
`setRequireStrongPassword()` true while we're on a member form is weird, but perfectly valid.

File: src/Forms/ConfirmedPasswordField.php [hunk 13]
```diff
@@ -723,6 +807,26 @@ public function getMinPasswordStrength(): int
         return $this->minPasswordStrength;
     }
 
+    /**
+     * Get the minimum password strength used for client-side evaluation
+     * where strength is defined on EntropyPasswordValidator
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]:
```suggestion
     * Get the minimum password strength used for client-side evaluation.
     * If we're on a member form, the strength may come from the member password validator.
```
It's not _only_ defined on that validator, because if `getRequireStrongPassword()` is true and `getMinPasswordStrength()` is higher than `$validator->getRequiredStrength()` (or we're not on a member form or the member uses a different validator), then we use `getRequireStrongPassword()`

File: src/Forms/ConfirmedPasswordField.php (hunk 5 above)
Comment (GuySartorelli) [reply src/Forms/ConfirmedPasswordField.php]: There is no relevant `PasswordValidator` if we're not on a member form, because that validator is explicitly only for member passwords.

File: src/Forms/ConfirmedPasswordField.php [hunk 14]
```diff
@@ -155,6 +177,47 @@ public function __construct(
...
+        $validator = $this->getApplicableValidator();
+        if (is_a($validator, EntropyPasswordValidator::class)) {
+            $strength = PasswordStrengthValidator::estimateStrength($password);
+            $strengthLabel = $validator->getStrengthLabel($strength);
+        }
+        // All translations are defined in PHP, rather than having a mixture
+        // of PHP and Javascript translations.
+        $message = _t(
+            __CLASS__ . '.STRENGTH',
+            'Password strength: {strengthLabel}.',
```
Comment (GuySartorelli) [inline src/Forms/ConfirmedPasswordField.php]:
```suggestion
            'Password strength: {strengthLabel}',
```
The dot looks weird

File: src/Security/MemberAuthenticator/ChangePasswordForm.php [hunk 15]
```diff
@@ -50,19 +53,15 @@ public function __construct($controller, $name, $fields = null, $actions = null)
      */
     protected function getFormFields()
     {
-        $fields = FieldList::create();
-
+        $field = ConfirmedPasswordField::create('Password');
```
Comment (GuySartorelli) [inline src/Security/MemberAuthenticator/ChangePasswordForm.php]: The labels shouldn't change, I think.
```suggestion
        $field = ConfirmedPasswordField::create('Password', _t(Member::class . '.NEWPASSWORD', 'New Password'), '', null, false, _t(Member::class . '.CONFIRMNEWPASSWORD', 'Confirm New Password'));
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Forms/ConfirmedPasswordField.php (hunk 9 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php:53]: I'd rather not

File: src/Forms/ConfirmedPasswordField.php (hunk 14 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: Updated

File: src/Forms/ConfirmedPasswordField.php (hunk 11 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php:255]: I think we still need the check about what the PasswordValidator is, if we're on a non-member form, getRequireStrongPassword() was called, however PasswordValidator configured is to RulesPasswordValidator, then we don't want to be making AJAX requests to the strength endpoint

File: src/Forms/ConfirmedPasswordField.php (hunk 10 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: Not changing - testing with `getMinPasswordStrengthForEvaluation()` looks really weird given that this is an endpoint

Also it'll mess up IDE type-hinting because the `if (is_a($validator, EntropyPasswordValidator::class)) {` makes the IDE know that `$validator->getStrengthLabel($strength);` is OK to call

File: src/Forms/ConfirmedPasswordField.php (hunk 12 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: Updated

File: src/Forms/ConfirmedPasswordField.php (hunk 13 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: Updated

File: src/Security/MemberAuthenticator/ChangePasswordForm.php (hunk 15 above)
Comment (emteknetnz) [reply src/Security/MemberAuthenticator/ChangePasswordForm.php]: Updated

File: src/Forms/ConfirmedPasswordField.php (hunk 5 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: I don't follow? There is a default class that Injector uses when instantiating PasswordValidators. That will be used by this field if isOnMemberForm is false and if requireStrongPassword has been set to true.

Also, I think at this point we're just splitting hairs, the exact internals of a few private methods that call each other don't really matter a whole lot

Comment (GuySartorelli) [reply src/Forms/ConfirmedPasswordField.php]: There's absolutely no reason to be referencing `PasswordValidator` _at all_ if we're not on a member form. That validator class and its subclasses are _only_ for member passwords. They aren't relevant at any other time. When `ConfirmedPasswordField` is used on the front-end or in some other arbitrary form, it doesn't care what the default password validator for members is.

I would agree that we're splitting hairs except that if someone sets this YAML config, they can't get an estimated strength on their perfectly valid `ConfirmedPasswordField::create()->setRequiresMinPasswordStrength(true)`:
```yml
---
After: '#corepasswords'
---
SilverStripe\Core\Injector\Injector:
  SilverStripe\Security\Validation\PasswordValidator:
    class: 'SilverStripe\Security\Validation\RulesPasswordValidator'
```

That config is what is used to update the password validator for members. It should _not_ affect the functionality for any `ConfirmedPasswordField` which is not on a member form.

File: src/Forms/ConfirmedPasswordField.php (hunk 10 above)
Comment (GuySartorelli) [reply src/Forms/ConfirmedPasswordField.php]: The current implementation is wrong. It will stop users from getting an esimation of their password strength for a valid `ConfirmedPasswordField::create()->setRequiresMinPasswordStrength(true)` if the _member_'s `PasswordValidator` is not an entropy password validator.

File: src/Security/MemberAuthenticator/ChangePasswordForm.php [hunk 16]
```diff
@@ -50,19 +53,20 @@ public function __construct($controller, $name, $fields = null, $actions = null)
      */
     protected function getFormFields()
     {
-        $fields = FieldList::create();
-
+        $field = ConfirmedPasswordField::create(
+            'Password',
+            _t(Member::class . '.NEWPASSWORD', 'New Password'),
+            titleConfirmField: _t(Member::class . '.CONFIRMNEWPASSWORD', 'Confirm New Password')
```
Comment (GuySartorelli) [inline src/Security/MemberAuthenticator/ChangePasswordForm.php]: You can't use named arguments with `create()`.

File: src/Security/MemberAuthenticator/ChangePasswordForm.php (hunk 15 above)
Comment (GuySartorelli) [reply src/Security/MemberAuthenticator/ChangePasswordForm.php]: Looks like you accidentally added two `$field = ` assignments.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

**Comment by GuySartorelli** [comment]
CI also still failing

File: src/Security/MemberAuthenticator/ChangePasswordForm.php (hunk 15 above)
Comment (emteknetnz) [reply src/Security/MemberAuthenticator/ChangePasswordForm.php]: Updated

File: src/Forms/ConfirmedPasswordField.php (hunk 10 above)
Comment (emteknetnz) [reply src/Forms/ConfirmedPasswordField.php]: Updated

File: src/Security/MemberAuthenticator/ChangePasswordForm.php (hunk 16 above)
Comment (emteknetnz) [reply src/Security/MemberAuthenticator/ChangePasswordForm.php]: Updated

File: composer.json [hunk 17]
```diff
@@ -54,7 +54,7 @@
         "symfony/mailer": "^7.0",
         "symfony/mime": "^7.0",
         "symfony/translation": "^7.0",
-        "symfony/validator": "^7.1",
+        "symfony/validator": "^7.2",
```
Comment (emteknetnz) [inline composer.json:57]: Needed because the visibility of [PasswordStrengthValidator::estimateStrength()](https://github.com/symfony/symfony/blob/7.2/src/Symfony/Component/Validator/Constraints/PasswordStrengthValidator.php#L60) changed from private to public in 7.2

File: tests/php/Security/Validation/EntropyPasswordValidatorTest.php [hunk 18]
```diff
@@ -6,6 +6,7 @@
 use SilverStripe\Dev\SapphireTest;
 use SilverStripe\Security\Member;
 use SilverStripe\Security\Validation\EntropyPasswordValidator;
+use Symfony\Component\Validator\Constraints\PasswordStrength;
```
Comment (GuySartorelli) [inline tests/php/Security/Validation/EntropyPasswordValidatorTest.php:9]: This isn't needed but it's not worth going through a whole review cycle for it - it's not doing any harm.
One day we will add a linting rule for unnecessary `use` statements :sweat_smile:

**Review by GuySartorelli — APPROVED** [review]
LGTM, MOG
